upgrade to webrtc m120-6099。

janreyho opened the pull request:
I've upgraded `libmediasoupclient` to WebRTC m120-6099 and also updated the corresponding test demo, `mediasoup-broadcaster-demo`, to the same version. 

This upgrade allows for successful room entry and video stream publishing. 

Testing on a Mac with an M3 Max chip has been successful.

janreyho committed:
upgrade to webrtc m120-6099。
You can successfully enter the room, publish the video stream。
On mac m3max test OK。

jmillan commented:
Hi @janreyho, 

The PR looks good, but the tests are still failing as in [PR](https://github.com/versatica/libmediasoupclient/tree/m120), which is what is blocking us from merging.

Would you please take a look at it so we make the test pass?

janreyho commented:
```python scripts/clang-tidy.py
/Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.clang-tidy:5:1: error: unknown key 'AnalyzeTemporaryDtors'
AnalyzeTemporaryDtors: false
^~~~~~~~~~~~~~~~~~~~~
Error parsing /Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.clang-tidy: Invalid argument
Enabled checks:
...
^~~~~~~~~~~~~~~~~~~~~
Error parsing /Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.clang-tidy: Invalid argument
/Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.clang-tidy:5:1: error: unknown key 'AnalyzeTemporaryDtors'
AnalyzeTemporaryDtors: false
^~~~~~~~~~~~~~~~~~~~~
Error parsing /Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.clang-tidy: Invalid argument

clang-tidy -header-filter=^/Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.* -p=/Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient /Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/src/DataConsumer.cpp

/Users/hejiayi/Desktop/8_haoxinyun/gitlab/hxSDK/mediasoup/libmediasoupclient/.clang-tidy:5:1: error: unknown key 'AnalyzeTemporaryDtors'
AnalyzeTemporaryDtors: false
```

python scripts/clang-tidy.py generates many above errors

I use [libmediasoupclient](https://github.com/janreyho/libmediasoupclient) in [mediasoup-broadcaster-demo](https://github.com/janreyho/mediasoup-broadcaster-demo) is OK, and  joining a room and publishing a stream is successfully。

jmillan commented:
@janreyho, as I commented in your PR the current problem is that `libmediasoupclient` tests are failing, so very likely is the tests that need changes.

janreyho commented:
```
sh scripts/test.sh
[INFO] compiling mediasoupclient and test_mediasoupclient: cmake --build build
[ 15%] Built target sdptransform
[ 76%] Built target mediasoupclient
[ 92%] Built target test_sdptransform
[100%] Built target sdptransform_readme_helper
[INFO] running tests: ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient
scripts/test.sh: line 45: ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient: No such file or directory
```

@jmillan how to generated test_mediasoupclient

jmillan commented:
In order to build the tests you need to set the `-DMEDIASOUPCLIENT_BUILD_TESTS:BOOLEAN=ON` flag.

```
cmake . -Bbuild \
  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
  -DMEDIASOUPCLIENT_BUILD_TESTS:BOOLEAN=ON \
  -DMEDIASOUPCLIENT_LOG_TRACE:BOOLEAN=ON \
  -DMEDIASOUPCLIENT_LOG_DEV:BOOLEAN=ON \
  -DLIBWEBRTC_INCLUDE_PATH:PATH=${PATH_TO_LIBWEBRTC_SOURCES} \
  -DLIBWEBRTC_BINARY_PATH:PATH=${PATH_TO_LIBWEBRTC_BINARY} \
  -DCMAKE_CXX_FLAGS="-fvisibility=hidden" \
  -DCLANG_TIDY_FIX=ON

make -C build/
```

janreyho committed:
The individual TEST_CASE tests have all passed, but the overall test reports：
All tests passed (338 assertions in 15 test cases)

[1]    36933 segmentation fault  ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient

janreyho commented:
@jmillan After modifying the three test case files:
```
test/src/Handler.test.cpp
test/src/MediaStreamTrackFactory.cpp
test/src/mediasoupclient.test.cpp
```
The individual TEST_CASE tests have all passed:
```assembly
(base) ➜  libmediasoupclient git:(m120) ✗ ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient mediasoupclient
Filters: mediasoupclient mediasoupclient
===============================================================================
All tests passed (189 assertions in 1 test case)

(base) ➜  libmediasoupclient git:(m120) ✗ ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient Handler
Filters: Handler Handler
===============================================================================
All tests passed (4 assertions in 1 test case)

(base) ➜  libmediasoupclient git:(m120) ✗
(base) ➜  libmediasoupclient git:(m120) ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient Device
Filters: Device Device
===============================================================================
All tests passed (15 assertions in 1 test case)

(base) ➜  libmediasoupclient git:(m120) ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient PeerConnection
Filters: PeerConnection PeerConnection
===============================================================================
All tests passed (8 assertions in 1 test case)

(base) ➜  libmediasoupclient git:(m120) ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient SendHandler
Filters: SendHandler SendHandler
===============================================================================
All tests passed (16 assertions in 1 test case)

(base) ➜  libmediasoupclient git:(m120)
```
Other test cases are also OK, but the overall test reports：segmentation fault
```assembly
(base) ➜  libmediasoupclient git:(m120) ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient
===============================================================================
All tests passed (338 assertions in 15 test cases)

[1]    45251 segmentation fault  ./build/test/test_mediasoupclient.app/Contents/MacOS/test_mediasoupclient
(base) ➜  libmediasoupclient git:(m120)
```
I still don't know how to locate the problem。

janreyho committed:
Create a singleton class for createFactory and ReleaseThreads

jmillan commented on test/include/MediaStreamTrackFactory.hpp:
```diff
@@ -2,6 +2,56 @@
+class Singleton {
+public:
+    static Singleton& getInstance() {
+        static Singleton instance; // 唯一实例
```
Please, comments only in english.

jmillan commented on test/include/MediaStreamTrackFactory.hpp:
```diff
@@ -2,6 +2,56 @@
+        return instance;
+    }
+
+    // 防止拷贝和赋值
```
Same, please replace all comments to english.

jmillan commented on test/include/MediaStreamTrackFactory.hpp:
```diff
@@ -2,6 +2,56 @@
+
+private:
+    Singleton() {
+        std::cout << "Singleton created\n";
```
remove.

jmillan commented on test/include/MediaStreamTrackFactory.hpp:
```diff
@@ -2,6 +2,56 @@
+        createFactory();
+    }
+    ~Singleton() {
+        std::cout << "Singleton destroyed\n";
```
remove.

jmillan commented on test/include/MediaStreamTrackFactory.hpp:
```diff
@@ -2,6 +2,56 @@
+        ReleaseThreads();
+    }
+
+	/* MediaStreamTrack holds reference to the threads of the PeerConnectionFactory.
+	* Use plain pointers in order to avoid threads being destructed before tracks.
+	*/
```
```suggestion
	/**
	* MediaStreamTrack holds reference to the threads of the PeerConnectionFactory.
	* Use plain pointers in order to avoid threads being destructed before tracks.
	*/
```

jmillan commented on test/src/Handler.test.cpp:
```diff
@@ -37,19 +37,20 @@ TEST_CASE("Handler", "[Handler]")
 
-	static std::unique_ptr<mediasoupclient::PeerConnection> pc(
-	  new mediasoupclient::PeerConnection(nullptr, &PeerConnectionOptions));
+	// static std::unique_ptr<mediasoupclient::PeerConnection> pc(
+	//   new mediasoupclient::PeerConnection(nullptr, &PeerConnectionOptions));
```
Just remove if not needed.

jmillan commented on test/src/MediaStreamTrackFactory.cpp:
```diff
@@ -1,3 +1,6 @@
+#ifndef SINGLETON_H
+#define SINGLETON_H
```
`MEDIA_STREAM_TRACK_FACTORY`

jmillan commented on test/src/MediaStreamTrackFactory.cpp:
```diff
@@ -9,80 +12,108 @@
-	WorkerThread->SetName("worker_thread", nullptr);
 
-	if (!NetworkThread->Start() || !SignalingThread->Start() || !WorkerThread->Start())
+	void Singleton::createFactory()
```
```suggestion
	void MediaStreamTrackFactory::create()
```

jmillan commented on test/src/MediaStreamTrackFactory.cpp:
```diff
@@ -9,80 +12,108 @@
 
-	webrtc::PeerConnectionInterface::RTCConfiguration config;
+	void Singleton::ReleaseThreads() {
+		// 调用 reset 函数以释放资源
```
englihs.

jmillan commented on test/src/MediaStreamTrackFactory.cpp:
```diff
@@ -9,80 +12,108 @@
+	void Singleton::ReleaseThreads() {
+		// 调用 reset 函数以释放资源
+		if (Factory) {
+			// Factory->Release();
```
remove code rather than comment.

jmillan commented on test/src/MediaStreamTrackFactory.cpp:
```diff
@@ -9,80 +12,108 @@
+			PeerConnectionOptions.factory = nullptr;
+			Factory = nullptr;
+		}
+		// if (NetworkThread) {
+		//     NetworkThread->Stop();
+		//     NetworkThread.reset();
+		// 	NetworkThread = nullptr;
+		// }
+
+		// if (WorkerThread) {
+		//     WorkerThread->Stop();
+		//     WorkerThread.reset();
+		// 	WorkerThread = nullptr;
+		// }
+
+		// if (SignalingThread) {
+		//     SignalingThread->Stop();
+		//     SignalingThread.reset();
+		// 	SignalingThread = nullptr;
+		// }
```
remove

jmillan commented on test/src/mediasoupclient.test.cpp:
```diff
@@ -809,5 +815,7 @@ TEST_CASE("mediasoupclient", "[mediasoupclient]")
 		REQUIRE_THROWS_AS(
 			sendTransport->UpdateIceServers(iceServers),
 			MediaSoupClientError);
+
```
remove

jmillan reviewed (approved):
Merging into versatica:m120. Comments will be addressed there.
